packages/ui-extensions-sdk/src/utils/security.ts

```diff
@@ -0,0 +1,113 @@
+import { RESOURCE_BATCH_INTERVAL } from '../constants';
+import { LoadedResourceMetaDataBatch } from '../types';
+
+import { setImmediateInterval } from './utils';
+
+export type LoadedResourceIds = Set<string>;
+
+const isPerformanceObjectSupported = () => {
+    return window.performance !== undefined && 'getEntries' in performance;
+};
+
+const isResourceTimingEntry = (
+    entry: PerformanceEntry
+): entry is PerformanceResourceTiming => {
+    return entry.entryType === 'resource';
+};
+
+const isRequestKind = (timing: PerformanceResourceTiming) => {
+    return (
+        timing.initiatorType === 'xmlhttprequest' ||
+        timing.initiatorType === 'fetch'
+    );
+};
+
+export const collectResourceUsage = (
+    previouslyLoadedResources: LoadedResourceIds
+): [LoadedResourceMetaDataBatch, LoadedResourceIds] => {
+    // Use resource timing api to to retrieve application loaded resources
+    const snapshot = performance.getEntriesByType('resource');
+
+    const resources = snapshot
+        .filter(isResourceTimingEntry)
+        // to avoid collecting the same resource twice, we use
+        // the Set difference: snapshot - previousSnapshot
+        .filter(r => !previouslyLoadedResources.has(r.name))
+        .map(r => {
+            // safe because r.name represents the resolved URL of the requested resource.
+            const urlHostname = new URL(r.name).hostname;
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏what browsers are you targeting? (`URL` is not available in IE)

Reply:
No IE 🙏